addons/sale/models/sale_order.py

```diff
@@ -1217,6 +1217,36 @@ def get_portal_last_transaction(self):
         self.ensure_one()
         return self.transaction_ids._get_last()
 
+    def _on_downpayment_creation(self):
+        if (
+                len(self.order_line.filtered(lambda line: line.is_downpayment)) == 1
+                and len(self.order_line.filtered(
+                    lambda line: line.display_type and line.name == _("Down Payments"))) == 0
+        ):
+            sequence_before_downpayment = max(
+                line.sequence for line in self.order_line if not line.is_downpayment
+            )
+
+            # create a section "Down Payment" and introduce the down payment after it
+            so_values = {
+                'name': _('Down Payments'),
+                'product_uom_qty': 0.0,
+                'order_id': self.id,
+                'display_type': 'line_section',
+                'sequence': 2 * sequence_before_downpayment,
+            }
+            so_line_obj = self.env['sale.order.line']
+            so_line_obj.create(so_values)
+
+            # alter the down payment to be after the rest of the down payments
+            down_payment = [line for line in self.order_line if line.is_downpayment][0]
+            down_payment._alter_sequence(2 * sequence_before_downpayment + 1)
```

Comment:
This whole logic should be moved in the invoicing wizard.
The down payment lines are only created through this wizard in the standard code, and the down payment section should be created together with the down payment line.